README.md
# JS Viewport Utils 3.0.1

Includes:

- A standalone, plain JavaScript object, `viewportUtils`
- AMD and Node/CommonJS support

## Installation

### AMD, Node.js, CommonJS

#### [NPM](https://www.npmjs.com/package/js-viewport-utils)

`npm install js-viewport-utils`

And then:

`var viewportUtils = require('js-viewport-utils');`

#### Bower:

`bower install js-viewport-utils`

### Traditional include

```js
<script src="path-to/js-viewport-utils.js"></script>
```

### In Angular projects / Typescript / ES6
```ts
import {inViewport, inViewportBottom} from "js-viewport-utils";
```

## Usage

### Basic, with optional settings

```js
var foo = document.getElementById("foo");
viewportUtils.inViewport(foo, {sides: "left"});
```

If working with jQuery

```js
var $foo = jQuery("#foo");
viewportUtils.inViewport($foo[0], {sides: "left"});
```

Sides can be passed as unique string with spaces

```js
var options = {sides: "top right left bottom"}
```

###Use another element as viewport

```js
var bar = document.querySelector("#bar"); // Alternative for `getElementById`
viewportUtils.inViewport(elem, {container: bar});
```

###Custom thresholds

```js
viewportUtils.inViewport(elem, {top: 12, right: 12});
```

For more options, see [Settings](#settings) section below.

### Shorthand notation

```js
viewportUtils.inViewportLeft(elem);
viewportUtils.inViewportTop(elem);
```

These will check for the matching side, with default settings

## Browser Support

- IE 8+
- All the others except Opera Mini
    + Tested in the latest stable Chrome, Firefox, Safari, and IE
    + No "new" JavaScript or quirky techniques are employed so it should work in all other modern browsers not specifically mentioned above

## License

MIT © [Tony Samperi](mailto:github@tonysamperi.it)

This library is free, open source, and GPL friendly. You can use it for
commercial projects, open source projects, or really almost whatever you want.

Attribution is required by MIT, SIL OLF, and CC BY licenses. Downloaded files already
contain embedded comments with sufficient
attribution, so you shouldn't need to do anything additional when using these
files normally.

## Author

Tony Samperi, [tonysamperi.github.io](https://tonysamperi.github.io/)


changelog.md
#1.0.0
* Release

#1.0.1
* Fixed prototypes
* Deleted index.js (useless)
* Added changelog
* Added readme and license

#1.0.2
* Fixed shorthands (for real)

#2.0.0
* Added typings
* New interfaces
* Improved exports (single methods instead of object)

#2.0.1
* Fixed issue when jQuery exists

#2.0.2
* Removed useless lodash dependency
* Fixed test page
* Added Jest tests
* Refactor

#2.0.3 (deprecated)
* Refactor

#2.0.4
* Fix bad config breaking umd behaviour (jsViewportUtils var not available anymore)

#3.0.0
* Removed jQuery support to optimize dependencies, if needed pass ```jqueryEl[0]```
* Updated source typescript version to 4.2.4

#3.0.1
* Version bump to push some dependabot fixes


test/index.js
(function () {
    const $test1 = document.getElementById("test1");
    const $test2 = document.getElementById("test2");
    const $test3 = document.getElementById("test3");
    const $test4 = document.getElementById("test4");
    const $viewport = document.getElementById("viewport");
//
    $test1.innerHTML = jsViewportUtils.inViewport(document.getElementById("blue_square"), {container: $viewport}) ? "YES" : "NO";
    $test2.innerHTML = jsViewportUtils.inViewportBottom(document.getElementById("blue_square"), {container: $viewport}) ? "YES" : "NO";
    $test3.innerHTML = jsViewportUtils.inViewport(document.getElementById("red_square"), {container: $viewport}) ? "YES" : "NO";
    $test4.innerHTML = jsViewportUtils.inViewport(document.getElementById("green_square"), {container: $viewport}) ? "YES" : "NO";
})();


test/js-viewport-utils.test.ts
import {JSDOM} from "jsdom";
import {inViewport, JsViewportSides} from "../src/js-viewport-utils";

const dom = new JSDOM();
(global as any).document = dom.window.document;
(global as any).window = dom.window;

const html = `<!DOCTYPE html>
<html lang="en">
  <head>
    <title>Test</title>
  </head>
  <body>
    <main id="viewport" style="overflow: visible; border: 2px dashed #BBB; padding: 16px; position: relative">
    <h2 style="margin: 0; color: #BBB">Viewport</h2>
        <div id="blue_square" style="width: 100px; height: 100px; background: blue;"></div>
        <div id="red_square" style="width: 100px; height: 100px; background: red; position: absolute; bottom: -150px"></div>
        <div id="green_square"
             style="width: 100px; height: 100px; background: green; position: absolute; left: calc(100% - 100px); bottom: -50px"></div>
    </main>
  </body>
</html>`;
document.body.innerHTML = html;

describe("js-viewport-utils test", () => {

    afterEach(() => {
        // console.info("CLEANING UP...");
        (global as any).window = void 0;
        // (global as any).document = void 0;
    });

    it("should return true (inViewport)", () => {
        const $viewport = document.getElementById("viewport") as HTMLElement;
        expect(inViewport(document.getElementById("blue_square") as HTMLElement, {container: $viewport})).toBeTruthy();
    });

    it("should return true (inViewport bottom)", () => {
        const $viewport = document.getElementById("viewport") as HTMLElement;
        const opts = {
            container: $viewport,
            sides: [JsViewportSides.BOTTOM]
        };

        expect(inViewport(document.getElementById("blue_square") as HTMLElement, opts)).toBeTruthy();
    });

    it("should return false (inViewport)", () => {
        const $viewport = document.getElementById("viewport") as HTMLElement;

        expect(inViewport(document.getElementById("red_square") as HTMLElement, {container: $viewport})).toBeTruthy();
    });

    it("should return false (inViewport)", () => {
        const $viewport = document.getElementById("viewport") as HTMLElement;

        expect(inViewport(document.getElementById("green_square") as HTMLElement, {container: $viewport})).toBeTruthy();
    });

});


rollup.config.js
import resolve from "rollup-plugin-node-resolve";
import commonjs from "rollup-plugin-commonjs";
import sourceMaps from "rollup-plugin-sourcemaps";
import typescript from "rollup-plugin-typescript2";
import json from "rollup-plugin-json";

const pkg = require("./package.json");
const libraryName = "js-viewport-utils";
const camelize = (str) => {
    return str.replace(/(?:^\w|[A-Z]|\b\w)/g, function (word, index) {
        return index === 0 ? word.toLowerCase() : word.toUpperCase();
    }).replace(/\s+/g, "");
};

export default {
    input: `src/${libraryName}.ts`,
    output: [
        {file: pkg.main, name: "jsViewportUtils", format: "umd", sourcemap: true},
        {file: pkg.module, format: "cjs", sourcemap: true}
    ],
    // Indicate here external modules you don't wanna include in your bundle (i.e.: 'lodash')
    external: [
        // "lodash.defaults",
        "jquery"
    ],
    watch: {
        include: "src/**"
    },
    plugins: [
        // Allow json resolution
        json(),
        // Compile TypeScript files
        typescript({useTsconfigDeclarationDir: true}),
        // Allow bundling cjs modules (unlike webpack, rollup doesn't understand cjs)
        commonjs(),
        // Allow node_modules resolution, so you can use 'external' to control
        // which external modules to include in the bundle
        // https://github.com/rollup/rollup-plugin-node-resolve#usage
        resolve(),
        // Resolve source maps to the original source
        sourceMaps()
    ]
};


rollup.config.ts
import resolve from "rollup-plugin-node-resolve";
import commonjs from "rollup-plugin-commonjs";
import sourceMaps from "rollup-plugin-sourcemaps";
import camelCase from "lodash.camelcase";
import typescript from "rollup-plugin-typescript2";
import json from "rollup-plugin-json";

const pkg = require("./package.json");

const libraryName = "js-viewport-utils";

export default {
    input: `src/${libraryName}.ts`,
    output: [
        {file: pkg.main, name: camelCase(libraryName), format: "umd", sourcemap: true},
        {file: pkg.module, format: "es", sourcemap: true},
    ],
    // Indicate here external modules you don't wanna include in your bundle (i.e.: 'lodash')
    external: [
        // "lodash.defaults",
        "jquery"
    ],
    watch: {
        include: "src/**"
    },
    plugins: [
        // Allow json resolution
        json(),
        // Compile TypeScript files
        typescript({useTsconfigDeclarationDir: true}),
        // Allow bundling cjs modules (unlike webpack, rollup doesn't understand cjs)
        commonjs(),
        // Allow node_modules resolution, so you can use 'external' to control
        // which external modules to include in the bundle
        // https://github.com/rollup/rollup-plugin-node-resolve#usage
        resolve(),
        // Resolve source maps to the original source
        sourceMaps()
    ]
};


src/js-viewport-utils.ts
/**
 * Viewport Utils
 *
 * @description A set of viewport utils
 * @author      Tony Samperi, tonysamperi.github.io
 * @version     3.0.1
 * @date        2023-03-09
 */

import {
    JsViewportSettings,
    JsViewportSides,
    JsViewportConfig,
    JsViewportSidesRef,
    jsViewportDefaults
} from "./js-viewport-utils.model";

export * from "./js-viewport-utils.model";

const jsViewportUtils = (() => {

    // IE 8 and lower fail this
    const canUseWindowDimensions: boolean = window && typeof window.innerHeight === typeof 0;

    const viewportUtils = {
        /**
         * Determines whether an element is within the viewport
         * @param  {HTMLElement}  elem       DOM Element (required)
         * @param  {JsViewportSettings}  options    Optional settings
         * @return {Boolean}            Whether the element was completely within the viewport
         */
        inViewport: (elem: HTMLElement, options?: JsViewportSettings) => {

            if (!elem || typeof elem !== "object" || (elem as HTMLElement).nodeType !== 1) {
                throw new Error("First argument must be an element!");
            }

            let containerBoundingRect: ClientRect | DOMRect;
            let containerScrollTop: number;
            let containerScrollLeft: number;
            let elemBoundingRect: ClientRect | DOMRect;
            let isWindow: boolean;
            const scrollBarWidths = [0, 0];
            const sidesRegex = /^TOP$|^RIGHT$|^BOTTOM$|^LEFT$|^ALL$/;
            // TMP VARS
            let i: number;
            let result: boolean = !1;

            // Build configuration from defaults and user-provided settings and metadata
            const config: JsViewportConfig = {
                ...jsViewportDefaults,
                ...options
            } as JsViewportConfig;

            // Use the window as the container if the user specified the body or a non-element
            if (config.container === document.body || (config.container as HTMLElement).nodeType !== 1) {
                config.container = window;
            }

            // Get the element's bounding rectangle with respect to the viewport
            elemBoundingRect = (elem as HTMLElement).getBoundingClientRect();

            isWindow = (config.container === window);

            const isSideIn: JsViewportSidesRef<() => boolean> = {
                [JsViewportSides.TOP]: () => {
                    if (isWindow) {
                        return (elemBoundingRect.top >= config.top);
                    }
                    else {
                        return (elemBoundingRect.top >= containerScrollTop - (containerScrollTop - containerBoundingRect.top) + config.top);
                    }
                },
                [JsViewportSides.RIGHT]: () => {
                    const ctScroll = isWindow ? containerScrollLeft : -scrollBarWidths[0];

                    return elemBoundingRect.right <= (containerBoundingRect.right + ctScroll - config.right);
                },
                [JsViewportSides.BOTTOM]: () => {
                    let containerHeight = 0;

                    if (isWindow) {
                        if (canUseWindowDimensions) {
                            containerHeight = (config.container as Window).innerHeight;
                        }
                        else if (document && document.documentElement) {
                            containerHeight = document.documentElement.clientHeight;
                        }
                    }
                    else {
                        containerHeight = containerBoundingRect.bottom;
                    }

                    return (elemBoundingRect.bottom <= containerHeight - scrollBarWidths[1] - config.bottom);
                },
                [JsViewportSides.LEFT]: () => {
                    if (isWindow) {
                        return (elemBoundingRect.left >= config.left);
                    }
                    else {
                        return (elemBoundingRect.left >= containerScrollLeft - (containerScrollLeft - containerBoundingRect.left) + config.left);
                    }
                },

                // Element is within all four boundaries
                [JsViewportSides.ALL]: () => {
                    return (isSideIn.TOP() && isSideIn.BOTTOM() && isSideIn.LEFT() && isSideIn.RIGHT());
                }
            };


            // Get viewport dimensions and offsets
            if (isWindow) {
                containerBoundingRect = document.documentElement.getBoundingClientRect();
                containerScrollTop = document.body.scrollTop;
                containerScrollLeft = window.scrollX || document.body.scrollLeft;
            }
            else {
                containerBoundingRect = (config.container as HTMLElement).getBoundingClientRect();
                containerScrollTop = (config.container as HTMLElement).scrollTop;
                containerScrollLeft = (config.container as HTMLElement).scrollLeft;
            }

            if (containerScrollLeft) {
                scrollBarWidths[0] = 18;
            }

            if (containerScrollTop) {
                scrollBarWidths[1] = 16;
            }


            // Loop through all of the sides
            i = config.sides.length;

            while (i--) {
                // Test the element against each side of the viewport that was requested
                if (sidesRegex.test(config.sides[i])) {
                    if (isSideIn[config.sides[i]]()) {
                        result = !0;
                    }
                    else {
                        result = !1;
                        break;
                    }
                }
            }

            return result;
        },
        inViewportTop: (element: HTMLElement) => {
            return viewportUtils.inViewport(element, {sides: [JsViewportSides.TOP]});
        },
        inViewportRight: (element: HTMLElement) => {
            return viewportUtils.inViewport(element, {sides: [JsViewportSides.RIGHT]});
        },
        inViewportBottom: (element: HTMLElement) => {
            return viewportUtils.inViewport(element, {sides: [JsViewportSides.BOTTOM]});
        },
        inViewportLeft: (element: HTMLElement) => {
            return viewportUtils.inViewport(element, {sides: [JsViewportSides.LEFT]});
        }
    };

    return viewportUtils;
})();

export const inViewport = jsViewportUtils.inViewport;
export const inViewportTop = jsViewportUtils.inViewportTop;
export const inViewportRight = jsViewportUtils.inViewportRight;
export const inViewportBottom = jsViewportUtils.inViewportBottom;
export const inViewportLeft = jsViewportUtils.inViewportLeft;


js-viewport-utils.js
/**
 * Viewport Utils
 *
 * @description A set of viewport utils
 * @author      Tony Samperi, tonysamperi.github.io
 * @version     1.0.0
 * @date        2018-10-08
 */
(function (root, name, factory) {
    // AMD
    if (typeof define === "function" && define.amd) {
        define([], factory);
    }
    // Node and CommonJS-like environments
    else if (typeof module !== "undefined" && typeof exports === "object") {
        module.exports = factory();
    }
    // Browser global
    else {
        root[name] = factory();
    }
}(this, "viewportUtils", function () {

    var canUseWindowDimensions = typeof window !== "undefined" && window.innerHeight !== undefined; // IE 8 and lower fail this

    var viewportUtils = {
        /**
         * Determines whether an element is within the viewport
         * @param  {Object}  elem       DOM Element (required)
         * @param  {Object}  options    Optional settings
         * @return {Boolean}            Whether the element was completely within the viewport
         */
        inViewport: function (elem, options) {

            var config = {};
            var containerScrollTop;
            var containerScrollLeft;
            var i;
            var isWindow;
            var scrollBarWidths = [0, 0];
            var settings;
            var sidesRegex = /^top$|^right$|^bottom$|^left$|^all$/;

            // Handle jQuery
            if (typeof jQuery !== "undefined" && elem instanceof jQuery) {
                elem = elem.get(0);
            }

            if (typeof elem !== "object" || elem.nodeType !== 1) {
                throw new Error("First argument must be an element");
            }

            settings = options || {};

            // Build configuration from defaults and user-provided settings and metadata
            config.container = settings.container || typeof document !== "undefined" ? document.body : window;
            config.sides = settings.sides || "all";
            config.top = settings.top || 0;
            config.right = settings.right || 0;
            config.bottom = settings.bottom || 0;
            config.left = settings.left || 0;

            // Extract the DOM node from a jQuery collection
            if (typeof jQuery !== "undefined" && config.container instanceof jQuery) {
                config.container = config.container.get(0);
            }

            // Use the window as the container if the user specified the body or a non-element
            if (config.container === document.body || config.container.nodeType !== 1) {
                config.container = window;
            }

            // Get the element's bounding rectangle with respect to the viewport
            elemBoundingRect = elem.getBoundingClientRect();

            isWindow = (config.container === window);

            isSideIn = {
                top: function () {
                    if (isWindow) {
                        return (elemBoundingRect.top >= config.top);
                    }
                    else {
                        return (elemBoundingRect.top >= containerScrollTop - (containerScrollTop - containerBoundingRect.top) + config.top);
                    }
                },
                right: function () {
                    if (isWindow) {
                        return (elemBoundingRect.right <= (containerBoundingRect.right + containerScrollLeft) - config.right);
                    }
                    else {
                        return (elemBoundingRect.right <= containerBoundingRect.right - scrollBarWidths[0] - config.right);
                    }
                },
                bottom: function () {
                    var containerHeight = 0;

                    if (isWindow) {
                        if (canUseWindowDimensions) {
                            containerHeight = config.container.innerHeight;
                        }
                        else if (document && document.documentElement) {
                            containerHeight = document.documentElement.clientHeight;
                        }
                    }
                    else {
                        containerHeight = containerBoundingRect.bottom;
                    }

                    return (elemBoundingRect.bottom <= containerHeight - scrollBarWidths[1] - config.bottom);
                },
                left: function () {
                    if (isWindow) {
                        return (elemBoundingRect.left >= config.left);
                    }
                    else {
                        return (elemBoundingRect.left >= containerScrollLeft - (containerScrollLeft - containerBoundingRect.left) + config.left);
                    }
                },

                // Element is within all four boundaries
                all: function () {
                    return (isSideIn.top() && isSideIn.bottom() && isSideIn.left() && isSideIn.right());
                }
            };


            // Get viewport dimensions and offsets
            if (isWindow) {
                containerBoundingRect = document.documentElement.getBoundingClientRect();
                containerScrollTop = document.body.scrollTop;
                containerScrollLeft = window.scrollX || document.body.scrollLeft;
            }
            else {
                containerBoundingRect = config.container.getBoundingClientRect();
                containerScrollTop = config.container.scrollTop;
                containerScrollLeft = config.container.scrollLeft;
            }

            if (containerScrollLeft) {
                scrollBarWidths[0] = 18;
            }

            if (containerScrollTop) {
                scrollBarWidths[1] = 16;
            }


            // Loop through all of the sides
            sides = config.sides.split(" ");
            i = sides.length;

            while (i--) {
                side = sides[i].toLowerCase();
                // Test the element against each side of the viewport that was requested
                if (sidesRegex.test(side)) {
                    if (isSideIn[side]()) {
                        result = true;
                    }
                    else {
                        result = false;
                        break;
                    }
                }
            }

            return result;
        },
        inViewportTop: function (element) {
            return viewportUtils.inViewport(element, {sides: "top"});
        },
        inViewportRight: function (element) {
            return viewportUtils.inViewport(element, {sides: "right"});
        },
        inViewportBottom: function (element) {
            return viewportUtils.inViewport(element, {sides: "bottom"});
        },
        inViewportLeft: function (element) {
            return viewportUtils.inViewport(element, {sides: "left"});
        }
    };

    return viewportUtils;
}));
